dev-docs/RFCs/v8.0/ui-module-rfc.md

```diff
@@ -0,0 +1,137 @@
+- Fullscreen: button to enter / exit fullscreen mode
+- Geolocate: button to track the user's location using the browser's Geolocation API
+- Legend: visual comparison of sizes, colors etc. corresponding to the rendered layers and viewport
+
+
+## Background
+
+Deck.gl does not currently export any DOM-based UI components. The documentation recommends react-map-gl as a companion library in the following pattern:
+
+```jsx
+import {StaticMap, MapContext, NavigationControl, FullscreenControl} from 'react-map-gl'
+
+<DeckGL
+  ContextProvider={MapContext.Provider}
+>
+  <StaticMap />
+  <NavigationControl />
+  <FullscreenControl />
+</DeckGL>
+```
+
+There are major limitations with this approach:
+
+- It is only supported by react-map-gl v5 and v6, versions no longer actively maintained. v7 dropped deck.gl support to prioritize standalone performance, see https://github.com/visgl/react-map-gl/issues/1646.
+- It pulls in an entire base map library even if the application does not need Mapbox.
+- It is React only, cannot be used in vanilla JS.
+- It is geospatial only, does not work with non-geospatial views.
+- It cannot be used via the declarative interface e.g. with pydeck.
+
+
+## Proposal
+
+1. Export a new interface `UIControl` from `@deck.gl/core`. It will define certain hook methods that a UI component can implement, along the lines of:
+
+```js
+interface UIControl<PropsT> {
+  /// Populated by core when mounted
+  deck?: Deck;
+  container?: HTMLDivElement;
+
+  constructor(props: PropsT);
+
+  /// Lifecycle hooks
+  onAdd: () => void;
+  onRemove: () => void;
+  setProps: (props: PropsT) => void;
+
+  /// Event hooks
+  onViewStateChange?: () => void;
+  onRedraw?: () => void;
+  onHover?: (info) => void;
+  onClick?: (info) => void;
+}
+```
+
+2. Add a new module `@deck.gl/ui` which exports common UI components that implement the `UIControl` interface.
+
+3. Add the following APIs to the `Deck` class:
+
+```js
+/// Prop type for React / declarative usage
+new Deck({
+  uiControls: UIControl[]
+})
+
+/// Imperative methods for vanilla JS usage
+addUIControl(control: UIControl): void
+removeUIControl(control: UIControl): void
+```
+
+Users will use this new module like the following:
+
+```js
+import {NavigationControl, FullscreenControl} from '@deck.gl/ui'
+
+new Deck({
+  ...
+  uiControls: [
+    new NavigationControl(),
+    new FullscreenControl()
+  ]
+})
+```
+
+## Implementation details and concerns
+
+### Terminology
+
+React and Mapbox users use "controls" to refer to UI components. However, we already call an input handler "controller" in the deck.gl context. I have seen users not differentiating the two in issues and discussions. I am concerned that introducing another concept called "control" is going to add to the confusion.
+
+### React
+
+All components in `@deck.gl/ui` should work without React. `@deck.gl/react` can offer thin wrappers that work with the `DeckGL` component:
```

Comment:
Nit: Wrappers or just JSX syntax like some existing constructs?

Any need to support refs and such things so that react apps can interact imperatively with these things?

Reply:
Some controls will benefit from ref / imperative APIs. For example the `UserLocateControl` in react-map-gl needs to support method `trigger()`. It will also be true for more complex controls, e.g. if Nebula implements this interface.